recipes/xkbcommon/all/conanfile.py

```diff
@@ -48,64 +51,87 @@ def config_options(self):
+            self.tool_requires("wayland-protocols/1.27")
 
     def layout(self):
         basic_layout(self, src_folder="src")
 
     def source(self):
-        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+        get(self, **self.conan_data["sources"][self.version],
+            destination=self.source_folder, strip_root=True)
 
     def generate(self):
         tc = MesonToolchain(self)
-
-        tc.project_options["libexecdir"] = "bin"
-        tc.project_options["libdir"] = "lib"
-        tc.project_options["datadir"] = "res"
-
         tc.project_options["enable-docs"] = False
         tc.project_options["enable-wayland"] = self.options.get_safe("with_wayland", False)
         tc.project_options["enable-x11"] = self.options.with_x11
-
         if self._has_xkbregistry_option:
             tc.project_options["enable-xkbregistry"] = self.options.xkbregistry
+        if self._has_build_profile:
+            tc.project_options["build.pkg_config_path"] = self.generators_folder
         tc.generate()
 
-    def build(self):
-        # Conan doesn't provide a `wayland-scanner.pc` file for the package in the _build_ context
-        meson_build_file = os.path.join(self.source_folder, "meson.build")
-        replace_in_file(self, meson_build_file,
-                        "wayland_scanner_dep = dependency('wayland-scanner', required: false, native: true)",
-                        "# wayland_scanner_dep = dependency('wayland-scanner', required: false, native: true)")
+        pkg_config_deps = PkgConfigDeps(self)
+        if self._has_build_profile and self.options.get_safe("with_wayland"):
+            pkg_config_deps.build_context_activated = ["wayland", "wayland-protocols"]
+            pkg_config_deps.build_context_suffix = {"wayland": "_BUILD"}
+        pkg_config_deps.generate()
 
-        replace_in_file(self, meson_build_file,
-                        "if not wayland_client_dep.found() or not wayland_protocols_dep.found() or not wayland_scanner_dep.found()",
-                        "if not wayland_client_dep.found() or not wayland_protocols_dep.found()")
+        virtual_build_env = VirtualBuildEnv(self)
+        virtual_build_env.generate()
 
-        replace_in_file(self, meson_build_file,
-                        "wayland_scanner = find_program(wayland_scanner_dep.get_pkgconfig_variable('wayland_scanner'))",
-                        "wayland_scanner = find_program('wayland-scanner')")
+    def build(self):
+        if self.options.get_safe("with_wayland"):
+            meson_build_file = os.path.join(self.source_folder, "meson.build")
+            if self._has_build_profile:
+                replace_in_file(self, meson_build_file,
+                                "wayland_scanner_dep = dependency('wayland-scanner', required: false, native: true)",
+                                "wayland_scanner_dep = dependency('wayland-scanner_BUILD', required: false, native: true)")
```

Comment:
I'll get an error if I don't set the `build_context_suffix` for the `wayland` package because this library requires the `wayland-client` library from the `wayland` package as a regular requires while simultaneously requiring the `wayland-scanner` tool from the `wayland` package as a `tool_requires`.

If we were able to place the build-context pkg-config files in a separate directory from the others, this would avoid naming conflicts and the necessary patching to find the `tool_requires` dependencies. Meson would be able to handle this gracefully just by setting the standard project option `build.pkg_config_path` to a value including the directory for the build-context pkg-config files.

Reply:
I've gone ahead and re-worked this PR to use a separate directory for the pkg-config files in the _build context_. This doesn't require using `replace_in_file` when a build profile is present.

Reply:
Sometimes I ask questions for myself to learn. Don't feel obligated to make changes the community (aka you) know more than me and that's a good thing.

Reading and looking qt the code I feel like it's a limitation on the client, like there's nothing in the layout for separation... seems like a useful feature


Maybe we can open an issue again the main repo and see what the team thinks?

Reply:
I created https://github.com/conan-io/conan/issues/12342 but feel free to correct me if I misunderstood or left something important out :)